packages/@glimmer/runtime/lib/destroyables.ts

```diff
@@ -0,0 +1,187 @@
+import { DEBUG } from '@glimmer/env';
+import { Destroyable, Destructor } from '@glimmer/interfaces';
+import { _WeakSet as WeakSet } from '@glimmer/util';
+
+const EAGER_DESTRUCTORS = new WeakMap<Destroyable, Destructor<Destroyable>[]>();
+const DESTRUCTORS = new WeakMap<Destroyable, Destructor<Destroyable>[]>();
+const DESTROYABLE_CHILDREN = new WeakMap<Destroyable, Destroyable[]>();
+const DESTROYABLE_PARENTS = new WeakMap<Destroyable, Destroyable[]>();
+
+export function associateDestroyableChild<T extends Destroyable>(parent: Destroyable, child: T): T {
+  if (DEBUG && isDestroying(parent)) {
+    throw new Error(
+      'Attempted to associate a destroyable child with an object that is already destroying or destroyed'
+    );
+  }
+
+  let children = DESTROYABLE_CHILDREN.get(parent);
+  let parents = DESTROYABLE_PARENTS.get(child);
+
+  if (children === undefined) {
+    children = [];
+    DESTROYABLE_CHILDREN.set(parent, children);
+  }
+
+  if (parents === undefined) {
+    parents = [];
+    DESTROYABLE_PARENTS.set(parent, parents);
```

Comment:
things can have multiple parents? is the idea that a child only lives as long as the shortest lived parent? if most things will only have one parent, this should use `Destroyable | Destroyable[]` as the type.

The fact that you always mutate and read parents and children together and you alway check if it has been registered further argues for a state object versus multiple weak collections, you would have one dictionary get or set then fast properties for the rest of the lookups and sets

Reply:
> is the idea that a child only lives as long as the shortest lived parent? 

Yes